### PR TITLE
Add advanced bifacial PV and ground irradiance modeling

### PR DESCRIPTION
**Pull request by Matthew-Boyd**
## Dependencies
[/Matthew-Boyd/wex:bifacial](https://github.com/Matthew-Boyd/wex/tree/bifacial)
[/Matthew-Boyd/ssc:bifacial](https://github.com/Matthew-Boyd/ssc/tree/bifacial)

## Description

Adds:
- Irradiance loss factor differentiation
- Rack shading loss
- Intra-module electrical mismatch
- Multi-albedo ground UI input
- Temporospatial ground and rear irradiance output
- Bifacial loss diagram additions

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist:

If you have added a new compute module in a SSC pull request related to this one, be sure to check the [Process Requirements](https://github.com/NREL/SAM/wiki/Compute-modules-in-SAM).

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deploy/runtime/ui/PV Albedo and Radiation.txt [hunk 1]
```diff
@@ -1,27 +1,27 @@
 PV Albedo and Radiation
 882
-215
-7
+639
```
Comment (mjprilliman) [inline deploy/runtime/ui/PV Albedo and Radiation.txt:3]: This UI form is not loading in correctly for me: 

image

**Review by mjprilliman — CHANGES_REQUESTED** [review]
Are there extra steps I need to take to get PV Albedo and Radiation.txt to work as a UI form? So far it is not being read properly.

File: deploy/runtime/ui/PV Losses.txt [hunk 2]
```diff
@@ -4743,19 +5548,25 @@ function pv_losses_enable_subarray_items( isub, en )
 }
 function bifacial_pv_losses_enable_subarray_items( isub, en)
 {
+	enable("calculate_rack_shading", false);	// keep false
+	enable("calculate_bifacial_electrical_mismatch", is_bifacial());
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply deploy/runtime/ui/PV Losses.txt:5552]: I thought it was cleaner with just the single 'Calculate' label at the top. @cpaulgilman ?

File: deploy/runtime/ui/PV Losses.txt [hunk 3]
```diff
@@ -4743,19 +5548,25 @@ function pv_losses_enable_subarray_items( isub, en )
 }
 function bifacial_pv_losses_enable_subarray_items( isub, en)
 {
+	enable("calculate_rack_shading", false);	// keep false
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply deploy/runtime/ui/PV Losses.txt:5551]: It's a planned feature but is taking longer than expected from our collaborator.

File: deploy/runtime/ui/PV Albedo and Radiation.txt (hunk 1 above)
Comment (Matthew-Boyd) [reply deploy/runtime/ui/PV Albedo and Radiation.txt:3]: > Are there extra steps I need to take to get PV Albedo and Radiation.txt to work as a UI form? So far it is not being read properly. I will review in more detail when I can get the albedo UI form working.

Yes, sorry, please see above dependencies (i.e., the bifacial branches of wex and ssc).

Comment (mjprilliman) [reply deploy/runtime/ui/PV Albedo and Radiation.txt:3]: Forgot about wex. Thanks!

**Review by mjprilliman — CHANGES_REQUESTED** [review]
- UI is a bit confusing because when weather file albedo checkbox is checked, spatial albedo uses weather file albedo and UI inputs in matrix do not apply. Is this intended behavior?
- Both spatial and single-value rear POA reported in outputs when spatial data is used. Single-value appears to be average of the spatial data.  Will users get confused on this? 
image

- Can these variable names indicate that they apply to the rear surface? 
image
-I guess it makes sense to average bifacial electrical mismatch over all subarrays (all in one output)?

This is what I found so far in SAM, still working through ssc.

**Comment by Matthew-Boyd** [comment]
> UI is a bit confusing because when weather file albedo checkbox is checked, spatial albedo uses weather file albedo and UI inputs in matrix do not apply. Is this intended behavior?

It is the intended behavior. If "Use albedo from weather file if available" is checked, then the (single) albedo from the weather file is used if present, on a timestep-by-timestep basis. If an albedo value is missing in the weather file for a given timestep, then the spatial albedo is used (if "Use spatial monthly albedo values" is selected). I could see this functionality being useful for parametric studies of different years or locations that include differences in snow cover or crop harvesting (which can also occur at a sub-monthly fidelity). This will be explained in the help, but if this feature is too confusing, it can just be removed. @janinefreeman ? @cpaulgilman ?

**Comment by Matthew-Boyd** [comment]
> The Lifetime Hourly output for the rear irradiance after reflection is calculated from the corresponding Matrix output, but I wouldn't think it any more confusing than both the Single and Annual energy values that SAM has for some outputs. It's also a final irradiance value after losses, so it's a metric used more than others for comparison and analysis, for example to the front total irradiance after reflection.  @cpaulgilman , what would you like with this?

**Comment by Matthew-Boyd** [comment]
> Can these variable names indicate that they apply to the rear surface?

Yes, done.

File: deploy/runtime/ui/PV Losses.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply deploy/runtime/ui/PV Losses.txt:5551]: I hid the 'Calculate?' label and checkbox for now.

File: deploy/runtime/ui/PV Losses.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply deploy/runtime/ui/PV Losses.txt:5552]: I added the 'Calculate?' label here as well.

**Comment by mjprilliman** [comment]
> > The Lifetime Hourly output for the rear irradiance after reflection is calculated from the corresponding Matrix output, but I wouldn't think it any more confusing than both the Single and Annual energy values that SAM has for some outputs. It's also a final irradiance value after losses, so it's a metric used more than others for comparison and analysis, for example to the front total irradiance after reflection.  @cpaulgilman , what would you like with this?

I'm fine with leaving the variables as they are just wanted to talk it through.

**Comment by cpaulgilman** [comment]
> > UI is a bit confusing because when weather file albedo checkbox is checked, spatial albedo uses weather file albedo and UI inputs in matrix do not apply. Is this intended behavior?
> 
> It is the intended behavior. If "Use albedo from weather file if available" is checked, then the (single) albedo from the weather file is used if present, on a timestep-by-timestep basis. If an albedo value is missing in the weather file for a given timestep, then the spatial albedo is used (if "Use spatial monthly albedo values" is selected). I could see this functionality being useful for parametric studies of different years or locations that include differences in snow cover or crop harvesting (which can also occur at a sub-monthly fidelity). This will be explained in the help, but if this feature is too confusing, it can just be removed. @janinefreeman ? @cpaulgilman ?

Let's keep this as is for now, and I will address this and some other albedo confusion in https://github.com/NREL/SAM/tree/pv-albedo-fix.

**Comment by cpaulgilman** [comment]
> > > The Lifetime Hourly output for the rear irradiance after reflection is calculated from the corresponding Matrix output, but I wouldn't think it any more confusing than both the Single and Annual energy values that SAM has for some outputs. It's also a final irradiance value after losses, so it's a metric used more than others for comparison and analysis, for example to the front total irradiance after reflection.  @cpaulgilman , what would you like with this?
> 
> I'm fine with leaving the variables as they are just wanted to talk it through.

Those "POA irradiance after..." outputs are confusing -- I think that should be addressed separately. (I think it might be clearer to show a single POA irradiance value and show the losses instead of a series of reduced POA values.)

**Review by mjprilliman — APPROVED** [review]

**Pull request merged**